### PR TITLE
feat: allow to search students by email, name and username

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR adds a new endpoint to search students by username, email or name.

### Testing instructions

1. Open the following endpoint in your browser:
  `http://local.overhang.io:8000/platform-plugin-communications/{course-id}/api/search_learners?query=edu&page=1&page_size=10`
2. Modify the query parameters and verify the API response matches them.
3. The response should look like the following:

```json
{
  "course_id": "course-v1:demo+demo+demo",
  "page": 1,
  "page_count": 1,
  "page_size": 50,
  "count": 2,
  "results": [
    {
      "username": "admin",
      "email": "admin@edunext.co",
      "name": "admin"
    },
    {
      "username": "student",
      "email": "student@edunext.co",
      "name": "student"
    }
  ]
}
```

### CODE REVIEW

File: platform_plugin_communications/api/views.py [hunk 1]
```diff
@@ -38,13 +43,26 @@
 @common_exceptions_400
 def send_email_api_view(request, course_id):
     """
-    Get the forum email notification preference for the user.
+    Sened email to learners api view.
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/api/views.py]:
```suggestion
    Send email to learners api view.
```

File: platform_plugin_communications/api/serializers.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+User = get_user_model()
+
+
+class UserSearchSerializer(serializers.ModelSerializer):
+    """
+    Serializer for User model.
+    """
+
+    name = serializers.CharField(source="profile.name", read_only=True)
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/api/serializers.py:15]: Can we send the name as in the Open edX APIs? I think it's inside the profile dict. Can we check how it's done?

File: platform_plugin_communications/api/views.py [hunk 3]
```diff
@@ -38,13 +43,26 @@
...
 
     Extracted from: lms.djangoapps.instructor.views.api import send_email
     """
     return send_email(request, course_id)
 
 
+@transaction.non_atomic_requests
+@require_GET
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/api/views.py]: why no cache? can't we use cache for the search query?

File: platform_plugin_communications/tests/test_views.py [hunk 4]
```diff
@@ -204,3 +205,109 @@ def test_send_email_to_individual_learners_with_schedule(
...
+    """
+    Test case for search_learners_api_view.
+    """
+
+    @patch("platform_plugin_communications.api.views.User")
+    def test_search_users(self, mock_User):
+        """
+        Test case for sending email to individual learners.
+        """
+
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/tests/test_views.py]:
```suggestion
```

File: platform_plugin_communications/tests/test_views.py [hunk 5]
```diff
@@ -204,3 +205,109 @@ def test_send_email_to_individual_learners_with_schedule(
...
+        ]
+        users_mock = []
+        for user in users:
+            user_mock = Mock()
+            user_mock.username = user["username"]
+            user_mock.email = user["email"]
+            user_mock.profile.name = user["name"]
+            users_mock.append(user_mock)
+        mock_User.objects.filter.return_value = users_mock
+        response = search_learner(request, course_id)
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/tests/test_views.py:241]:
```suggestion

        response = search_learner(request, course_id)
```

File: platform_plugin_communications/tests/test_views.py [hunk 6]
```diff
@@ -204,3 +205,109 @@ def test_send_email_to_individual_learners_with_schedule(
...
+
+        mock_User.objects.filter.assert_called_once_with(
+            is_active=True,
+            courseenrollment__course_id=course_key,
+            courseenrollment__is_active=True,
+        )
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/tests/test_views.py:247]: can we also check this filter? https://github.com/eduNEXT/platform-plugin-communications/pull/4/files#diff-13fbc08c712019edc8bfb712dff48e3f4c70405f0ea8b7da2902a3d8812c31d3R142-R147

File: platform_plugin_communications/tests/test_views.py [hunk 7]
```diff
@@ -204,3 +205,109 @@ def test_send_email_to_individual_learners_with_schedule(
...
+            "course_id": course_id,
+            "results": users,
+        }
+
+    @patch("platform_plugin_communications.api.views.User")
+    def test_search_users_with_query(self, mock_User):
+        """
+        Test case for sending email to individual learners.
+        """
+
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/tests/test_views.py]:
```suggestion
```

File: platform_plugin_communications/tests/test_views.py (hunk 6 above)
Comment (mariajgrimaldi) [reply platform_plugin_communications/tests/test_views.py:247]: what's the difference between this test and the next one?

File: platform_plugin_communications/tests/test_views.py [hunk 8]
```diff
@@ -204,3 +205,109 @@ def test_send_email_to_individual_learners_with_schedule(
...
+            user_mock.username = user["username"]
+            user_mock.email = user["email"]
+            user_mock.profile.name = user["name"]
+            users_mock.append(user_mock)
+        mock_filter = Mock()
+        mock_User.objects.filter.return_value = mock_filter
+        mock_filter2 = Mock()
+        mock_filter.filter.return_value = mock_filter2
+        mock_filter2.distinct.return_value = users_mock
+        response = search_learner(request, course_id)
```
Comment (mariajgrimaldi) [inline platform_plugin_communications/tests/test_views.py:293]:
```suggestion

        response = search_learner(request, course_id)
```

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

File: platform_plugin_communications/api/serializers.py (hunk 2 above)
Comment (Ian2012) [reply platform_plugin_communications/api/serializers.py:15]: It's stored in the profile.name field: 
image

Comment (mariajgrimaldi) [reply platform_plugin_communications/api/serializers.py:15]: how is it returned by the users API? Like this?
```
{
"username": "admin",
"profile": {
  "name": "Administrator"
 }
}
```

Comment (Ian2012) [reply platform_plugin_communications/api/serializers.py:15]: It uses the `name`:

image

File: platform_plugin_communications/tests/test_views.py (hunk 6 above)
Comment (Ian2012) [reply platform_plugin_communications/tests/test_views.py:247]: The next one covers the case where the query param is defined. In that case, the queryset is updated

Comment (Ian2012) [reply platform_plugin_communications/tests/test_views.py:247]: > can we also check this filter? https://github.com/eduNEXT/platform-plugin-communications/pull/4/files#diff-13fbc08c712019edc8bfb712dff48e3f4c70405f0ea8b7da2902a3d8812c31d3R142-R147

This test case doesn't trigger that behavior. The following test triggers it

**Comment by Ian2012** [comment]
@mariajgrimaldi added a new cache on the queryset itself and removed the cache_control decorator

File: platform_plugin_communications/tests/test_views.py (hunk 6 above)
Comment (mariajgrimaldi) [reply platform_plugin_communications/tests/test_views.py:247]: Can we update the docstring and the test name so it reflects that?

**Review by mariajgrimaldi — APPROVED** [review]
It's working perfectly! Thanks

**Pull request merged**